packages/codegen/src/codegen.ts

```diff
@@ -0,0 +1,316 @@
+      .reduce((acc, path) => {
+        return Object.assign(acc, { [path]: this.versionFile.paths[path] });
+      }, {} as OpenApiPathsObject);
+  };
+
+  private copyResponsesToSchemas = (): void => {
+    if (typeof this.versionFile.components!.responses === 'undefined') {
+      return;
+    }
+
+    // some refs are responses, and we must fetch out the schema for any json responses
+    for (const [name, response] of Object.entries(
+      this.versionFile.components!.responses
+    )) {
+      if (isReferenceObject(response)) {
+        continue;
+      }
+
+      const schema = response.content?.['application/json']?.schema;
+      if (typeof schema === 'undefined') {
+        continue;
+      }
+
+      const missingResponseSuffix = !name.toLowerCase().endsWith('response');
+      const schemaName = missingResponseSuffix ? name + 'Response' : name;
+      this.versionFile.components!.schemas![schemaName] = schema;
+    }
+  };
+
+  private queryParameters = () => {
+    const parameters = this.versionFile.components?.parameters;
+    if (typeof parameters === 'undefined') {
+      return {};
+    }
+
+    const schemas: Record<string, any> = {};
+    for (const [name, parameter] of Object.entries(parameters)) {
+      if (isReferenceObject(parameter)) {
+        continue;
+      }
+
+      if (parameter.in != 'query') {
+        continue;
+      }
+
+      const schema = {
+        type: 'object',
+        properties: Object.assign({}, { [parameter.name]: parameter.schema }),
+      };
+      schemas[name] = schema;
+    }
+
+    return schemas;
+  };
+
+  private copyQueryParametersToSchemas = (): void => {
+    const parameters = this.queryParameters();
+    for (const [name, schema] of Object.entries(parameters)) {
+      this.versionFile.components!.schemas![name + 'QueryParameter'] = schema;
+    }
+  };
+
+  public generateTypesFromSchemas = async (
+    schemaFilter?: StringFilter
+  ): Promise<string[]> => {
+    if (typeof this.versionFile.components?.schemas === 'undefined') {
+      throw new Error('components.schema is undefined');
+    }
+
+    if (typeof schemaFilter !== 'undefined') {
+      const schemas = this.versionFile.components!.schemas!;
+      this.versionFile.components!.schemas = Object.keys(schemas)
+        .filter(schemaFilter)
+        .reduce(
+          (acc, schemaName) =>
+            Object.assign(acc, { [schemaName]: schemas[schemaName] }),
+          {}
+        );
+    }
+
+    const typeNames = Object.keys(this.versionFile.components?.schemas || {});
+    await this.runSwaggerGenerator();
+
+    return typeNames;
+  };
+
+  /**
+   * generateTypes
+   *
+   * generate all types with a given versionfile and configuration. Returns name of all generated types afterwards.
+   */
+  public generateTypes = async (): Promise<string[]> => {
+    this.filterPaths();
+    this.removeSchemasWithoutPathRelations();
+    this.copyResponsesToSchemas();
+    this.copyQueryParametersToSchemas();
```

Comment:
I'm not a large fan of code like this. It makes the flow of operations and effects harder to follow, and e.g. breaks in the case of method being called twice and such, since it's building state in the object.

I'd recommend being explicit about the values you pass around, as in removing `this.versionFile` at least. I'm neutral on doing mutations on the values in the methods though - that should be fine as long as we are working on a deep copy.

Reply:
As an example here it was tricky to understand that the `this.versionFile` used in `runSwaggerGenerator` is _not_ the same as what the object was initialized with. I'd say that in the context of `runSwaggerGenerator` that variables has a bad and confusing name. It could be a lot more clear if the signature was something like `private runSwaggerGenerator = async (transformedSpec: OpenApiDocument) =>` instead

I also think it would be more clear if the side effects (such as writing `types.gen.ts`) was not hidden so deep and mixed with other operations. It's also confusing in regard to `generateTypesFromSchemas` vs `generateTypes`. Invoking only the first will not do what the second is doing, but both end up writing a types file (in the same place with different content?).

Reply:
`generateTypesFromSchemas` will only look at the schemas. While `generateTypes` will try to derive more types than just those defined in `schemas`.

Reply:
I don't think schemas are altered in `generateTypes`, only new schemas are added (unless unused ones are filtered out).